lib/uklock/include/uk/mutex.h

```diff
@@ -98,53 +100,71 @@ void uk_mutex_get_metrics(struct uk_mutex_metrics *dst);
 static inline void uk_mutex_lock(struct uk_mutex *m)
 {
 	struct uk_thread *current;
-	unsigned long irqf;
 
 	UK_ASSERT(m);
 
 	current = uk_thread_current();
 
+	/* if the owner is current thread then it definitely can't unlock during
+	 * uk_mutex_lock, hence if the owner is current thread then we can simply
+	 * increment the lock count
+	 */
+#ifdef CONFIG_LIBUKLOCK_MUTEX_RECURSE
+	if (unlikely(m->owner == current)) {
+		ukarch_inc(&(m->lock_count));
+		return;
+	}
+#else
+	UK_ASSERT(m->owner != current);
+#endif
+
+
 	for (;;) {
-		uk_waitq_wait_event(&m->wait,
-			m->lock_count == 0 || m->owner == current);
-		irqf = ukplat_lcpu_save_irqf();
-		if (m->lock_count == 0 || m->owner == current)
+		/* TODO: make wait queue SMP safe */
+		uk_waitq_wait_event(&m->wait, m->owner == NULL);
+
+		/* If there is no owner for the lock we can grab it and increment the
+		 * lock count
+		 *
+		 * Here the owner act as the lock
+		 * Here We dont need to disable interrupts because the compare exchange
+		 * is atomic even on the single processor
+		 * TODO: discuss the possibility that interrupt handler may require the
+		 * same lock, resulting in deadlock.
```

Comment:
My gut feeling here is that disabling (the) IRQ should be done by the user if needed and not generally in every case.

But if you have something like:
```C
flags = ukplat_lcpu_save_irqf();
uk_mutex_lock(&mutex);
/* do something */
uk_mutex_unlock(&mutex);
ukplat_lcpu_restore_irqf(flags);
```
and you don't get the mutex and would block, it is really bad that IRQs are turned off as long as you block (which might actually also deadlock the system).

Nevertheless, acquiring a mutex in an IRQ handler is not a great idea anyways and the proper way would be to use a spinlock with `uk_spin_lock_irqf()` instead.

Reply:
Hi @marcrittinghaus, you are right that it is the programmer's job not to sleep while holding a lock or not hold a lock which is needed by an interrupt handler.